File: sellia-app/src/app/core/services/cash-operation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface CashOperation {
  publicId: string;
  cashierSessionPublicId: string;
  cashierName: string;
  userName: string;
  type: 'ENTREE' | 'SORTIE';
  amount: number;
  description: string;
  reference?: string;
  authorizedBy: string;
  operationDate: string;
  adminNotes?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CashOperationTotals {
  totalEntrees: number;
  totalSorties: number;
  netAmount: number;
  entreesCount: number;
  sortiesCount: number;
}

export interface CashOperationCreateRequest {
  cashierSessionId: string;
  type: 'ENTREE' | 'SORTIE';
  amount: number;
  description: string;
  reference?: string;
  authorizedBy: string;
}

@Injectable({
  providedIn: 'root'
})
export class CashOperationService {
  private apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  createCashOperation(request: CashOperationCreateRequest): Observable<CashOperation> {
    return this.http.post<CashOperation>(`${this.apiUrl}/cash-operations`, request);
  }

  getOperationsBySession(sessionId: string): Observable<CashOperation[]> {
    return this.http.get<CashOperation[]>(`${this.apiUrl}/cash-operations/session/${sessionId}`);
  }

  getOperationsBySessionPaged(sessionId: string, page: number = 0, size: number = 20): Observable<any> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/cash-operations/session/${sessionId}/paged`, { params });
  }

  getTotalsBySession(sessionId: string): Observable<CashOperationTotals> {
    return this.http.get<CashOperationTotals>(`${this.apiUrl}/cash-operations/session/${sessionId}/totals`);
  }

  getAllOperations(page: number = 0, size: number = 20): Observable<any> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/cash-operations`, { params });
  }

  getOperationsByDateRange(startDate: string, endDate: string, page: number = 0, size: number = 20): Observable<any> {
    const params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/cash-operations/date-range`, { params });
  }

  getOperationsByCashier(cashierId: string, page: number = 0, size: number = 20): Observable<any> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<any>(`${this.apiUrl}/cash-operations/cashier/${cashierId}`, { params });
  }

  getOperationById(publicId: string): Observable<CashOperation> {
    return this.http.get<CashOperation>(`${this.apiUrl}/cash-operations/${publicId}`);
  }

  updateAdminNotes(publicId: string, adminNotes: string): Observable<CashOperation> {
    return this.http.put<CashOperation>(`${this.apiUrl}/cash-operations/${publicId}/admin-notes`, { adminNotes });
  }

  deleteCashOperation(publicId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/cash-operations/${publicId}`);
  }
}

File: sellia-backend/src/main/java/com/follysitou/sellia_backend/service/CashOperationService.java
package com.follysitou.sellia_backend.service;

import com.follysitou.sellia_backend.dto.request.CashOperationCreateRequest;
import com.follysitou.sellia_backend.dto.request.CashOperationUpdateAdminNotesRequest;
import com.follysitou.sellia_backend.dto.response.CashOperationResponse;
import com.follysitou.sellia_backend.dto.response.CashOperationTotalsResponse;
import com.follysitou.sellia_backend.enums.CashOperationType;
import com.follysitou.sellia_backend.exception.ResourceNotFoundException;
import com.follysitou.sellia_backend.mapper.CashOperationMapper;
import com.follysitou.sellia_backend.model.CashOperation;
import com.follysitou.sellia_backend.model.CashierSession;
import com.follysitou.sellia_backend.model.User;
import com.follysitou.sellia_backend.repository.CashOperationRepository;
import com.follysitou.sellia_backend.repository.CashierSessionRepository;
import com.follysitou.sellia_backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CashOperationService {

    private final CashOperationRepository cashOperationRepository;
    private final CashOperationMapper cashOperationMapper;
    private final CashierSessionRepository cashierSessionRepository;
    private final UserRepository userRepository;

    @Transactional
    public CashOperationResponse createCashOperation(CashOperationCreateRequest request) {
        // Get cashier session
        CashierSession session = cashierSessionRepository.findByPublicId(request.getCashierSessionId())
                .orElseThrow(() -> new ResourceNotFoundException("CashierSession", "publicId", request.getCashierSessionId()));

        // Get current user
        User currentUser = getCurrentUser();

        // Create cash operation
        CashOperation cashOperation = CashOperation.builder()
                .cashierSession(session)
                .user(currentUser)
                .type(request.getType())
                .amount(request.getAmount())
                .description(request.getDescription())
                .reference(request.getReference())
                .authorizedBy(request.getAuthorizedBy())
                .operationDate(LocalDateTime.now())
                .build();

        CashOperation saved = cashOperationRepository.save(cashOperation);
        return cashOperationMapper.toResponse(saved);
    }

    @Transactional
    public CashOperationResponse updateAdminNotes(String publicId, CashOperationUpdateAdminNotesRequest request) {
        CashOperation cashOperation = getCashOperationEntityById(publicId);
        cashOperation.setAdminNotes(request.getAdminNotes());
        CashOperation updated = cashOperationRepository.save(cashOperation);
        return cashOperationMapper.toResponse(updated);
    }

    public List<CashOperationResponse> getOperationsBySession(String sessionId) {
        List<CashOperation> operations = cashOperationRepository.findByCashierSession(sessionId);
        return operations.stream()
                .map(cashOperationMapper::toResponse)
                .collect(Collectors.toList());
    }

    public Page<CashOperationResponse> getOperationsBySessionPaged(String sessionId, Pageable pageable) {
        Page<CashOperation> operations = cashOperationRepository.findByCashierSessionPaged(sessionId, pageable);
        return operations.map(cashOperationMapper::toResponse);
    }

    public Page<CashOperationResponse> getAllOperations(Pageable pageable) {
        Page<CashOperation> operations = cashOperationRepository.findAllActive(pageable);
        return operations.map(cashOperationMapper::toResponse);
    }

    public Page<CashOperationResponse> getOperationsByDateRange(LocalDateTime startDate, LocalDateTime endDate, Pageable pageable) {
        Page<CashOperation> operations = cashOperationRepository.findByDateRange(startDate, endDate, pageable);
        return operations.map(cashOperationMapper::toResponse);
    }

    public Page<CashOperationResponse> getOperationsByCashier(String cashierId, Pageable pageable) {
        Page<CashOperation> operations = cashOperationRepository.findByCashier(cashierId, pageable);
        return operations.map(cashOperationMapper::toResponse);
    }

    public CashOperationResponse getOperationById(String publicId) {
        CashOperation cashOperation = getCashOperationEntityById(publicId);
        return cashOperationMapper.toResponse(cashOperation);
    }

    public CashOperationTotalsResponse getTotalsBySession(String sessionId) {
        Long totalEntrees = cashOperationRepository.getTotalBySessionAndType(sessionId, CashOperationType.ENTREE);
        Long totalSorties = cashOperationRepository.getTotalBySessionAndType(sessionId, CashOperationType.SORTIE);
        Integer entreesCount = cashOperationRepository.getCountBySessionAndType(sessionId, CashOperationType.ENTREE);
        Integer sortiesCount = cashOperationRepository.getCountBySessionAndType(sessionId, CashOperationType.SORTIE);

        Long netAmount = totalEntrees - totalSorties;

        return new CashOperationTotalsResponse(
                totalEntrees,
                totalSorties,
                netAmount,
                entreesCount,
                sortiesCount
        );
    }

    @Transactional
    public void deleteCashOperation(String publicId) {
        CashOperation cashOperation = getCashOperationEntityById(publicId);
        cashOperation.setDeleted(true);
        cashOperationRepository.save(cashOperation);
    }

    private CashOperation getCashOperationEntityById(String publicId) {
        return cashOperationRepository.findByPublicId(publicId)
                .orElseThrow(() -> new ResourceNotFoundException("CashOperation", "publicId", publicId));
    }

    private User getCurrentUser() {
        String userId = com.follysitou.sellia_backend.util.SecurityUtil.getCurrentUserId();
        if (userId == null) {
            throw new com.follysitou.sellia_backend.exception.UnauthorizedException("User is not authenticated");
        }
        return userRepository.findByPublicId(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "publicId", userId));
    }
}
